### PR TITLE
Don't static link external C libs

### PR DESCRIPTION
**Pull request by hjiawei**
## Description

This changeset disable external glibc static linking and disable cgo as much as possible when boringssl or external C libs are not required.

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
TBD
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -101,7 +101,7 @@ e2e-test:
 ###############################################################################
 # Build the release tool.
 hack/release/release: $(shell find ./hack/release -type f -name '*.go')
-	$(DOCKER_RUN) $(CALICO_BUILD) go build -v -o $@ ./hack/release/cmd
+	$(call build_binary, ./hack/release/cmd, $@)
```
Comment (hjiawei) [inline Makefile:104]: Always disable cgo for binaries that are not shipped in production images.

File: apiserver/Makefile [hunk 2]
```diff
@@ -99,8 +99,7 @@ else
 	$(call build_binary, $(PACKAGE_NAME)/cmd/apiserver, $@)
 endif
 else
-	$(DOCKER_RUN) -e CGO_ENABLED=$(CGO_ENABLED) $(CALICO_BUILD) \
-		sh -c '$(GIT_CONFIG_SSH) go build -v -o $@ -ldflags "$(VERSION_FLAGS)" $(PACKAGE_NAME)/cmd/apiserver'
+	$(call build_binary, $(PACKAGE_NAME)/cmd/apiserver, $@)
```
Comment (hjiawei) [inline apiserver/Makefile:102]: Always disable cgo for arches that don't support fips/boringssl expect for binaries that explicitly require cgo.

File: apiserver/Makefile [hunk 3]
```diff
@@ -137,11 +136,6 @@ $(CONTAINER_FIPS_CREATED): docker-image/Dockerfile.$(ARCH) $(BINDIR)/apiserver-$
 lint-cache-dir:
 	mkdir -p $(CURDIR)/.lint-cache
 
-check-boring-ssl: $(BINDIR)/apiserver-$(ARCH)
```
Comment (hjiawei) [inline apiserver/Makefile:140]: This is part of `build_cgo_boring_binary` helper function in lib.Makefile. No need a separate target to validate.

File: calicoctl/Makefile [hunk 4]
```diff
@@ -76,11 +76,7 @@ bin/calicoctl-linux-%: BUILDOS=linux
 bin/calicoctl-%: $(SRC_FILES)
 	$(MAKE) build-calicoctl BUILDOS=$(BUILDOS) ARCH=$(ARCH)
 build-calicoctl:
-ifeq ($(BUILDOS)/$(ARCH), linux/amd64)
-	$(call build_static_cgo_boring_binary, ./calicoctl/calicoctl.go, bin/calicoctl-$(BUILDOS)-$(ARCH))
-else
 	$(call build_binary, ./calicoctl/calicoctl.go, bin/calicoctl-$(BUILDOS)-$(ARCH))
```
Comment (hjiawei) [inline calicoctl/Makefile:79]: Always disable cgo for binaries that runs as a client from the client machine.

File: cni-plugin/Makefile [hunk 5]
```diff
@@ -95,22 +90,20 @@ sub-build-%:
...
 $(WINDOWS_BIN)/install.exe: $(SRC_FILES)
-	-mkdir -p $(WINDOWS_BIN)
-	$(DOCKER_RUN) \
-	-e GOOS=windows \
-	-e GOARCH=amd64 \
-	-e CGO_ENABLED=1 \
-		$(CALICO_BUILD) sh -c '$(GIT_CONFIG_SSH) \
-		go build -v -buildvcs=false -o $(WINDOWS_BIN)/install.exe -ldflags "$(LDFLAGS)" $(PACKAGE_NAME)/cmd/calico'
+	mkdir -p $(WINDOWS_BIN)
+	$(call build_windows_binary, $(PACKAGE_NAME)/cmd/calico, $@)
```
Comment (hjiawei) [inline cni-plugin/Makefile:106]: Always disable cgo for windows binaries.

File: pod2daemon/Makefile [hunk 6]
```diff
@@ -77,35 +77,22 @@ bin/flexvol-armv7: ARCH=armv7
...
 $(BINDIR)/csi-driver-%: $(SRC_FILES)
-	$(DOCKER_RUN) -e CGO_ENABLED=$(CGO_ENABLED) $(CALICO_BUILD) go build -buildvcs=false -v -o $(BINDIR)/csi-driver-$(ARCH) csidriver/main.go
-
-$(BINDIR)/csi-driver-amd64: $(SRC_FILES)
-ifeq ($(FIPS), true)
-	$(call build_static_cgo_boring_binary, csidriver/main.go, $@)
-else
-	$(call build_static_binary, csidriver/main.go, $@)
-endif
+	$(call build_binary, csidriver/main.go, $@)
```
Comment (hjiawei) [inline pod2daemon/Makefile:87]: I believe this csi driver talks to kubelet via unix domain socket and no encryption is involved so cgo is disabled.

File: lib.Makefile [hunk 7]
```diff
@@ -190,54 +169,43 @@ endef
...
 # For binaries that do not require boring crypto.
 define build_binary
-	$(DOCKER_RUN) $(CALICO_BUILD) \
-		sh -c '$(GIT_CONFIG_SSH) \
-		go build -o $(2)  \
-		-v -buildvcs=false \
-		-ldflags "$(LDFLAGS)" \
-		$(1)'
+	$(DOCKER_RUN) \
+		-e CGO_ENABLED=0 \
```
Comment (hjiawei) [inline lib.Makefile:194]: `build_binary` now always build a cgo disabled binary.

File: lib.Makefile [hunk 8]
```diff
@@ -190,54 +169,43 @@ endef
...
-        $(DOCKER_RUN) $(CALICO_BUILD) \
-                sh -c '$(GIT_CONFIG_SSH) \
-                go build -o $(2)  \
-                -v -buildvcs=false \
-                -ldflags "$(LDFLAGS) -linkmode external -extldflags -static" \
-                $(1)'
+# For windows builds that do not require cgo.
+define build_windows_binary
+	$(DOCKER_RUN) \
+		-e CGO_ENABLED=0 \
```
Comment (hjiawei) [inline lib.Makefile:202]: cgo is always disabled for windows binaries and arch is set to `amd64`.

File: app-policy/Makefile [hunk 9]
```diff
@@ -123,15 +113,12 @@ $(BINDIR)/healthz-s390x: ARCH=s390x
 $(BINDIR)/healthz-%: protobuf $(SRC_FILES)
 ifeq ($(ARCH),amd64)
 ifeq ($(FIPS), true)
-	$(call build_static_cgo_boring_binary, ./cmd/healthz, $@)
+	$(call build_cgo_boring_binary, ./cmd/healthz, $@)
 else
-	$(call build_static_binary, ./cmd/healthz, $@)
+	$(call build_binary, ./cmd/healthz, $@)
```
[the remark this answers is not in the document]
Comment (hjiawei) [reply app-policy/Makefile:113]: done.

File: apiserver/Makefile (hunk 2 above)
Comment (hjiawei) [reply apiserver/Makefile]: done.

File: app-policy/Makefile [hunk 10]
```diff
@@ -110,10 +103,7 @@ else
 	$(call build_binary, ./cmd/dikastes, $@)
 endif
 else
-	$(DOCKER_RUN_RO) \
-	  -e CGO_ENABLED=$(CGO_ENABLED) \
-	  -v $(CURDIR)/bin:/go/src/$(PACKAGE_NAME)/bin \
-	  $(CALICO_BUILD) go build -buildvcs=false -ldflags "-X main.VERSION=$(GIT_VERSION)" -v -o $(BINDIR)/dikastes-$(ARCH) ./cmd/dikastes
+	$(call build_binary, ./cmd/dikastes, $@)
```
[the remark this answers is not in the document]
Comment (hjiawei) [reply app-policy/Makefile]: done.

**Review by rene-dekker — APPROVED** [review]
lgtm

**Pull request merged**